File: py/Clonegraph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        
        if node == None:
            return None;
        
        head = UndirectedGraphNode(node.label);
        dict = {head.label:head}
        
        queue = []
            
        queue.insert(0,node)
        
        while queue != []:
            tmpnode = queue.pop();
            
            for x in tmpnode.neighbors:
                if dict.has_key(x.label) == False:
                    p = UndirectedGraphNode(x.label);
                    queue.insert(0,x);
                    dict[x.label] = p;
                    dict[tmpnode.label].neighbors.append(p);
                else:
                    dict[tmpnode.label].neighbors.append(dict[x.label]);
        
        return head

File: cpp/digit_problem/Clone Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        map<int ,int> labelmap;
        map<int,UndirectedGraphNode*> nodemap;
        
        
        queue<UndirectedGraphNode*> qgraph;
        
        if(node == NULL)
            return NULL;
        
        qgraph.push(node);
        UndirectedGraphNode *newhead = new UndirectedGraphNode(node->label);
        
        nodemap[node->label] = newhead;
        labelmap[node->label] = 1;
        
        
        while(qgraph.size() != 0)
        {
            UndirectedGraphNode* tmpnode = qgraph.front();
            qgraph.pop();
            
            for(int i = 0;i < tmpnode->neighbors.size();i++)
            {
                if(labelmap.find(tmpnode->neighbors[i]->label) == labelmap.end())
                {
                    labelmap[tmpnode->neighbors[i]->label] = 1;
                    qgraph.push(tmpnode->neighbors[i]);
                }
                
                if(nodemap.find(tmpnode->neighbors[i]->label) == nodemap.end())
                {
                    UndirectedGraphNode* newnode = new UndirectedGraphNode(tmpnode->neighbors[i]->label);
                    nodemap[tmpnode->neighbors[i]->label] = newnode;
                    
                    nodemap[tmpnode->label]->neighbors.push_back(newnode);
                }
                else
                {
                    nodemap[tmpnode->label]->neighbors.push_back(nodemap[tmpnode->neighbors[i]->label]);
                }
            }
        }
        
        return newhead;
    }
};